docs/developers/README.md

```diff
@@ -536,14 +537,44 @@ integration tests run. When the flag in your cluster is `alpha` _all_
 integration tests are executed, both `stable` and `alpha`. Setting the feature
 flag to `stable` will exclude `alpha` tests.
 
-## What and Why of `/tekton/steps`
+## What and Why of `/tekton/run`
 
-`/tekton/steps/` is an implicit volume mounted on a pod and created for storing
-the step specific information/metadata. There is one more subdirectory created
-under `/tekton/steps/` for each step in a task.
+`/tekton/run` is a collection of implicit volumes mounted on a pod and created
+for storing the step specific information/metadata. Steps can only write
+metadata to their own `/run` directory - all other step volumes are mounted as
+`readonly`.
 
-Let's take an example of a task with three steps, each exiting with non-zero
-exit code:
+### `/tekton/steps`
+
+`/tekton/steps` contains Step information that is intended to be used by user
+steps (and is therefore covered by the Tekton API compatibility policy).
```

Comment:
I'm wondering:
1. intended to be used by user steps in what way?
2. this is currently not included in our compat policy, maybe we need to update it - our compatibility policy currently mentions "the structure of the directories created in executing containers by Tekton" (https://github.com/tektoncd/pipeline/blob/main/api_compatibility_policy.md#what-is-the-api) and links to https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#reserved-directories which indicates everything other than /workspace and /tekton/results is considered an implementation detail

Reply:
i think im generally not understanding why we want both steps and run to exist - i think i understand that "steps" is intended to be used by the contents of the steps themselves (the user specified steps) but i dont understand why - and also if run is readonly anyway (except for the current step) im not sure why we need the 2 dirs

Reply:
We have conflicting documentation - https://github.com/tektoncd/pipeline/blob/5dcd43431fe1fdc62a4ed59fde3f01ca43536d72/docs/developers/README.md?plain=1#L195-L200 says it is in scope.

If it's not in scope, I'm more than happy to remove the `/steps` directory altogether. :D The symlink shenanigans are purely to preserve the /tekton/steps file structure for backwards compatibility.

Reply:
/cc @pritidesai who wrote the original documentation in https://github.com/tektoncd/pipeline/commit/92746a2362117c729ad865717dd0dab5ec3d326b

Reply:
ah yikes! some scope creep but if we could fix this so that we are only attempting to list the folders in the compat policy in one place that would be ideal 🙏  (pretty good chance adding 2 sources was my doing...)

Looking at TEP-0040 and the origin of the change, it's not clear to me if that folder needs to be part of the API spec or not (https://github.com/tektoncd/community/blob/main/teps/0040-ignore-step-errors.md#proposal), since we provide `$(steps.<step-name>.exitCode.path)` that does give us some freedom to potentially change the location of the directory. The TEP even says:

> If you would like to use the tekton internal path, you can access the exit code by reading the file (it is not recommended though)

i.e. the path is referred to as "internal" several times in the TEP

so basically I have mixed feelings - if @pritidesai and/or others feel strongly the path should remain part of our API, no objections from me - but if possible I'd rather consider that an implementation detail and reply on variable replacement insteaed

Reply:
thanks @wlynch and @bobcatfish for bringing my attention to this!

TEP-0040 proposed using the path variable `$(steps.<step-name>.exitCode.path)` and avoid using the tekton internal path. 

At the same time, we had a vision of having the `debug` feature utilizing the `/steps` directory structure (not yet implemented) and the symlinks will make it easier to invoke the debug scripts.

https://github.com/tektoncd/community/pull/463#issuecomment-873225465 

Just based on TEP-0040, I would avoid specifying it as covered by the Tekton API compatibility policy.

Reply:
SGTM! I'll make these changes in another PR so it's clear why we're making the change.

Reply:
Done in #4395. To limit the scope of this PR, I'm keeping the /tekton/steps folder (since I'm not sure what else is depending on it and how intrusive of a change it would be to remove). We can look into removing this in another PR.

Reply:
Tiny nit to close the loop here: remove the remaining parenthetical in this commit describing /tekton/steps as covered ("and is therefore covered by the Tekton API compatibility policy").

Reply:
Done.